BungeeCord-Patches/0060-Additional-DoS-mitigations.patch

```diff
@@ -49,6 +55,129 @@ index d10cf2ed..31a95495 100644
++
++    // Cached Exceptions:
++    private static final CorruptedFrameException PACKET_LENGTH_OVERSIZED =
++            new CorruptedFrameException("A packet could not be framed because it was too large. For more "
++                    + "information, launch Waterfall with -Dwaterfall.packet-decode-logging=true");
++
++    private static Exception handleOverflow(int packetCount, int expected, int actual) {
++        if (DEBUG) {
++            throw new CorruptedFrameException( "Packet #" + packetCount + " could not be framed because was too large" +
++                    " (expected " + expected + " bytes, got " + actual + " bytes)");
++        } else {
++            return PACKET_LENGTH_OVERSIZED;
++        }
++    }
++    // Waterfall end
++
++
++    @Override
++    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception
++    {
++        // If we decode an invalid packet and an exception is thrown (thus triggering a close of the connection),
++        // the Netty ByteToMessageDecoder will continue to frame more packets and potentially call fireChannelRead()
++        // on them, likely with more invalid packets. Therefore, check if the connection is no longer active and if so
++        // sliently discard the packet.
++        if ( !ctx.channel().isActive() )
++        {
++            in.skipBytes( in.readableBytes() );
++            return;
++        }
++
++        in.markReaderIndex();
++
++        for ( int i = 0; i < 3; i++ ) // Waterfall
++        {
++            if ( !in.isReadable() )
++            {
++                in.resetReaderIndex();
++                return;
++            }
++
++            // Waterfall start
++            byte read = in.readByte();
++            if ( read >= 0 )
++            {
++                in.resetReaderIndex();
++                int length = DefinedPacket.readVarInt( in );
++                // Waterfall end
++                if ( false && length == 0) // Waterfall - ignore
++                {
++                    throw new CorruptedFrameException( "Empty Packet!" );
++                }
++                // Waterfall start - length sanity checks for first packets in client connection
++                if (packetCount < 4) {
++                    // in case we get server list ping, the the packets we get after the first one are always smaller
++                    // than any of these, so no need for special casing
++                    int maxLength = 2097151; // max length of 21-bit varint
++                    switch (packetCount) {
++                        case 0:
++                            maxLength = Handshake.EXPECTED_MAX_LENGTH + 2;
++                            break;
++                        case 1:
++                            maxLength = LoginRequest.EXPECTED_MAX_LENGTH + 1;
++                            break;
++                        case 2:
++                            maxLength = EncryptionResponse.EXPECTED_LENGTH + 2;
```

Comment:
offline mode does not use encryprion stuff.

Reply:
Oh yeah forgot about that. Will add a check for that.

Reply:
I've put in higher limits to account for offline mode.